docs: add blog post for overrides (#4689)
aws/copilot-cli: site/content/blogs/release-v127.en.md

## Extend Copilot-generated AWS CloudFormation templates

AWS Copilot enables builders to quickly get started with containerized applications through the `copilot init` command and following prompts.
Developers can then grow their applications by editing and deploying [manifest](../docs/manifest/overview.en.md) infrastructure-as-code files.
And now with v1.27, developers can extend any Copilot-generated CloudFormation template with the `copilot [noun] override` command, enabling them to fully customize their infrastructure.

### AWS Cloud Development Kit (CDK) Overrides

You can use the CDK to extend your CloudFormation templates when you need the expressive power and safety of a
programming language. After running the `copilot [noun] override` command, Copilot will generate a CDK application
under the `copilot/[name]/override` directory:

```console
.
├── bin/
│   └── override.ts
├── .gitignore
├── cdk.json
├── package.json
├── README.md
├── stack.ts
└── tsconfig.json
```

> 

Reviewer: Do we want to point users to `README.md` as well, for more instructions?

        public readonly template: cdk.cloudformation_include.CfnInclude;
        public readonly appName: string;
        public readonly envName: string;
    
        constructor (scope: cdk.App, id: string, props: TransformedStackProps) {
            super(scope, id, props);
            this.template = new cdk.cloudformation_include.CfnInclude(this, 'Template', {
                templateFile: path.join('.build', 'in.yml'),
            });
            this.appName = props.appName;
            this.envName = props.envName;
            this.transformPublicNetworkLoadBalancer();
        }
    
        /**
         * transformPublicNetworkLoadBalancer removes the "Subnets" properties from the NLB,
         * and adds a SubnetMappings with predefined elastic IP addresses.
         */
        transformPublicNetworkLoadBalancer() {
            const elasticIPs = [new ec2.CfnEIP(this, 'ElasticIP1'), new ec2.CfnEIP(this, 'ElasticIP2')];
            const publicSubnets = cdk.Fn.importValue(`${this.appName}-${this.envName}-PublicSubnets`);
    
            // Apply the override.
            const nlb = this.template.getResource("PublicNetworkLoadBalancer") as elbv2.CfnLoadBalancer;

>             nlb.addDeletionOverride('Properties.Subnets');

Reviewer 2: Does it make sense to link https://docs.aws.amazon.com/cdk/v2/guide/cfn_layer.html or similar somewhere in the docs to help them better write their override code?

            nlb.subnetMappings = [{
                allocationId: elasticIPs[0].attrAllocationId,
                subnetId: cdk.Fn.select(0, cdk.Fn.split(",", publicSubnets)),
            }, {
                allocationId: elasticIPs[1].attrAllocationId,
                subnetId: cdk.Fn.select(1, cdk.Fn.split(",", publicSubnets)),
            }]
        }
    }
    ```

To learn more about extending with the CDK, checkout the [guide](../docs/developing/overrides/cdk.md).

### YAML Patch Overrides

> You can use YAML Patch overrides for a more lightweight experience when you do not want to have a dependency

Reviewer 3: ```suggestion
You can use YAML Patch overrides for a more lightweight experience when 1) you do not want to have a dependency
```

> You can use YAML Patch overrides for a more lightweight experience when you do not want to have a dependency
> on any other tooling and framework, or 2) you have to write only a handful modifications. 

Reviewer 2: ```suggestion
You can use YAML Patch overrides for a more lightweight experience when 1) you do not want to have a dependency
on any other tooling and framework, or 2) you have to write only a handful modifications. 
```

> on any other tooling and framework, or 2) you have to write only a handful modifications. 

Reviewer 4: ```suggestion
on any other tooling and framework, or 2) you have to write only a handful of modifications. 
```

After running the `copilot [noun] override` command, Copilot will generate a sample `cfn.patches.yml` file
under the `copilot/[name]/override` directory:

```console
.
├── cfn.patches.yml
└── README.md
```

> You can get started with adding, removing, or replacing properties by editing the `cfn.patchs.yaml` file.

Reviewer 3: ```suggestion
You can get started with adding, removing, or replacing properties by editing the `cfn.patches.yaml` file.
```

??? note "View sample `cfn.patches.yml`"

    ```yaml
    - op: add
      path: /Mappings
      value:
        ContainerSettings:
          test: { Cpu: 256, Mem: 512 }
          prod: { Cpu: 1024, Mem: 1024}
    - op: remove
      path: /Resources/TaskRole
    - op: replace
      path: /Resources/TaskDefinition/Properties/ContainerDefinitions/1/Essential
      value: false
    - op: add
      path: /Resources/Service/Properties/ServiceConnectConfiguration/Services/0/ClientAliases/-
      value:
        Port: !Ref TargetPort
        DnsName: yamlpatchiscool
    ```

> 

Reviewer 4: Do we need to add that they should run `copilot svc deploy` after this?

> To learn more about extending with YAML patches, checkout the [guide](../docs/developing/overrides/yamlpatch.md).

Reviewer 4: ```suggestion
To learn more about extending with YAML patches, check out the [guide](../docs/developing/overrides/yamlpatch.md).
```

## Preview AWS CloudFormation template changes

##### `copilot [noun] package --diff`

You can now run `copilot [noun] package --diff` to see the diff between your local changes and the latest deployed template.
The program will exit after it prints the diff.

!!! info "The exit codes when using `copilot [noun] package --diff`"

    0 = no diffs found  
    1 = diffs found  

>     2 = error producing diffs

Reviewer 4: ```suggestion
    2 = error-producing diffs
```

##### `copilot [noun] deploy --diff`

Similar to `copilot [noun] package --diff`, you can run `copilot [noun] deploy --diff` to see the same diff.

> However, instead of exiting after it print the diff, Copilot will follow up with a question: `Continue with the deployment? [y/N]`.

Reviewer 4: ```suggestion
However, instead of exiting after printing the diff, Copilot will follow up with the question: `Continue with the deployment? [y/N]`.
```
